### DISCUSSION TITLE
Is there some additional documentation on Magical Types?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#7972] [Q&A]
I could only find this, which seems to be unfinished: https://psalm.dev/docs/annotating_code/type_syntax/atomic_types/#magical-types

I'm looking for something similar to mapped types, because I'm trying to remove multiple keys from a shape. unset() is able to do that, but so far I'm unable to replicate the behaviour.

To clarify, I need either something like typescript's Omit<>, or some way of subtracting a shape from another shape.

Thanks.

**Comment by orklah** [comment]
I don't know Typescript, so I may be completely off in my answer.

In Psalm (and also in PHPStan), array shapes are not "strict". Which means that any key not directly documented in the shape could still be in the array without issue. (which means, if you document a function as expection `array{foo: string}` you can pass an `array{foo: string, bar: string}` without getting an error.

We don't currently have a strict array (which would forbid other keys in the shape). The closest thing we do have is that you should be able to document a known forbidden key as `never` which should refuse any type

**Reply by rzvc** [reply]
I'm aware about that, but you wouldn't be able to pass `array{bar: string}` to it, so while shapes are not sealed, there's still checking going on, and I'm trying to make use of that behaviour.

**Reply by orklah** [reply]
I'm not sure in what context you are. If you have a plugin and you know the type of the variable, you can directly change the type of the variable. If you're not in a plugin, can you explain more?

**Reply by rzvc** [reply]
Sure, consider the following code:
```php
function consumeFoo(&$data)
{
	unset($data['foo']);
}

function consumeBar(&$data)
{
	unset($data['bar']);
}

$data = ['foo' => '', 'bar' => ''];
```
How can I type this, so the following happens:
```php
consumeFoo($data);	// OK. $data is now array{bar: string}
consumeBar($data);	// OK. $data is now array{}
consumeFoo($data);	// Not OK.
```

**Reply by AndrolGenhald** [reply]
@orklah typescript has a bunch of useful utility types like [`Partial`](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype) that create new types based on existing types, where (using psalm syntax instead of typescript) `Partial<array{foo: int, bar: string}>` would be equivalent to `array{foo?: int, bar?: string}`. [`Omit`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) would remove a key from a type, so `Omit<array{foo: int, bar: string}, "foo">` would be equivalent to `array{bar: string}`.

@rzvc I don't think we currently support anything like this, but it would be neat to try at some point.

**Reply by orklah** [reply]
Ok, clearly it can't be done with vanilla Psalm. So you'd have to go to plugins.

Now two things:
- Either you know the behaviour of consumeFoo and you can hardcode it in your plugin
- Or you want to create a generic feature

For the first one, it's easier. You can just check for any call to consumeFoo, retrieve the type, and if it's a shape, just dump the `foo` property.

For the second one, you'll probably have to design your own personnal phpdoc tag, parse it and dump the documented property

**Reply by rzvc** [reply]
@AndrolGenhald, from what I'm observing, the functionality is already there, because unset() actually has that effect on the shape, but I believe it's something tucked in in the UnsetAnalyzer, and it might not be exposed to anything else yet.

https://psalm.dev/r/6d9e35273f

**Reply by AndrolGenhald** [reply]
@rzvc Yes, UnsetAnalyzer modifies the type of the variable directly. There's no way to annotate that though. It would be interesting to try adding something like `Omit<...>`, but with the way the type system currently works it's probably not feasible. I think it might work a bit better once/if I ever finish my type system rewrite.
